src/BenchmarksApps/HttpClientBenchmarks/Clients/HttpClient/Program.cs

```diff
@@ -0,0 +1,321 @@
+        BenchmarksEventSource.Register("http/exceptions", Operations.Sum, Operations.Sum, "Exceptions", "Number of exceptions", "n0");
+        BenchmarksEventSource.Register("http/rps/mean", Operations.Avg, Operations.Avg, "Mean RPS", "Requests per second - mean", "n0");
+
+        if (s_options.CollectRequestTimings)
+        {
+            RegisterPercentiledMetric("http/headers", "Time to headers (ms)", "Time to headers (ms)");
+            RegisterPercentiledMetric("http/contentstart", "Time to first content byte (ms)", "Time to first content byte (ms)");
+            RegisterPercentiledMetric("http/contentend", "Time to last content byte (ms)", "Time to last content byte (ms)");
+        }
+
+        Func<HttpMessageInvoker, Task<Metrics>> scenario;
+        switch(s_options.Scenario)
+        {
+            case "get":
+                scenario = Get;
+                break;
+            default:
+                throw new ArgumentException($"Unknown scenario: {s_options.Scenario}");
+        }
+
+        s_isWarmup = true;
+        s_isRunning = true;
+
+        var coordinatorTask = Task.Run(async () =>
+        {
+            await Task.Delay(TimeSpan.FromSeconds(s_options.Warmup));
+            s_isWarmup = false;
+            Log("Completing warmup...");
+            await Task.Delay(TimeSpan.FromSeconds(s_options.Duration));
+            s_isRunning = false;
+            Log("Completing scenario...");
+        });
+
+        var tasks = new List<Task<Metrics>>(s_options.NumberOfHttpClients * s_options.ConcurrencyPerHttpClient);
+        for (int i = 0; i < s_options.NumberOfHttpClients; ++i)
+        {
+            var client = s_httpClients[i];
+            for (int j = 0; j < s_options.ConcurrencyPerHttpClient; ++j)
+            {
+                tasks.Add(scenario(client));
+            }
+        }
+
+        await coordinatorTask;
+        var metricsArray = await Task.WhenAll(tasks);
+        foreach (var metrics in metricsArray)
+        {
+            s_metrics.Add(metrics);
+        }
+
+        LogMetric("http/successrequests", s_metrics.SuccessRequests);
+        LogMetric("http/badstatusrequests", s_metrics.BadStatusRequests);
+        LogMetric("http/exceptions", s_metrics.ExceptionRequests);
+        LogMetric("http/rps/mean", s_metrics.MeanRps);
+
+        if (s_options.CollectRequestTimings && s_metrics.SuccessRequests > 0)
+        {
+            LogPercentiledMetric("http/headers", s_metrics.HeadersTimes, TicksToMs);
+            LogPercentiledMetric("http/contentstart", s_metrics.ContentStartTimes, TicksToMs);
+            LogPercentiledMetric("http/contentend", s_metrics.ContentEndTimes, TicksToMs);
+        }
+    }
+
+    private static async Task<Metrics> Get(HttpMessageInvoker client)
+    {
+        var uri = new Uri(s_options.Url + "/get");
+
+        var drainArray = new byte[81920];
```

Comment:
Is this allocating for each request?

Reply:
No, it's per thread (the method contains a loop of requests below)